File: 2020/D19/d19.py
"""
Advent Of Code 2020
Day 19: xxx part 1 & 2

https://adventofcode.com/2020/day/19
"""
from time import perf_counter as pfc
import os

def readInput(filename):
  with open(os.path.join("2020", "D19", filename)) as f:
    lines = f.readlines()
  rules = {}
  linesIter = iter(lines)
  for l in linesIter:
    if not l.strip('\n'):
      break
    rid, rule = l.strip("\n").split(':')
    rid = int(rid)
    rules[rid] = []
    if "a" in rule:
      rules[rid].append(True)
      rules[rid].append(tuple(tuple('a')))
      continue
    if "b" in rule:
      rules[rid].append(True)
      rules[rid].append(tuple(tuple('b')))
      continue
    rules[rid].append(False)
    t = []
    for r in rule.split("|"):
      t.append("".join(tuple(f"<{id}>" for id in r.strip(' \n').split(' '))))
    rules[rid].append(t)
  messages = list(l.strip(' \n') for l in linesIter)
  messages.sort()
  # messages = tuple(m for m in messages if len(m) == 24)
  return rules, tuple(messages)

def reduceRule(crid, seq, rule):
  p = f"<{crid}>"
  reduced = []
  for r in rule:
    n = r.count(p)
    if n == 0:
      reduced.append(r)
    elif n > 1 and len(seq) > 1:
      reduced.extend(r.replace(p, e) for e in seq)
      # print(f"{crid} : {rule} - {seq}")
      # nn = [r] * n * len(seq)
      # rCount = 0
      # for i, n in enumerate(nn):
      #   nn[i] = n.replace(p, seq[rCount], 1)

      # nn = [r.replace(p, e, i + 1) for i in range(n) for e in seq]
      # reduced.extend(n1.replace(p, e) for n1 in nn for e in seq)
      # print(f"{reduced}")
    else:
      reduced.extend(r.replace(p, e) for e in seq)
  return reduced

def part1(rules, messages):
  startTime = pfc()
  #reduce rules
  aRules = []
  for r in range(len(rules)):
    aRules.append(rules[r])

  while not aRules[0][0]:
    # find a completed rule
    for crid, rule in enumerate(aRules):
      if not rule:
        continue
      if rule[0]:
        break
    seq = rule[1]
    # replace rule id
    for rid, rule in enumerate(aRules):
      if not rule:
        continue
      if rule[0]:
        continue
      ne = reduceRule(crid, seq, rule[1])
      completed = all(map(lambda x: x.isalpha(), ne))
      aRules[rid] = (completed, ne)
    aRules[crid] = None

  rules = aRules[0][1]
  m = tuple(m for m in messages if len(m) == 24 and m in rules)
  m = set(m)
  r = set(rules).intersection(m)
  result = len(r)

  print(f"Part 1 time: {pfc() - startTime:.4}")
  return result

def part2(input):
  startTime = pfc()
  print(f"Part 2 time: {pfc() - startTime:.4}")
  return

rules, messages = readInput(f"d19-input.txt")
#rules, messages = readInput(f"d19-sample.txt")
print(f"Part 1 for puzzle input: {part1(rules, messages)}")
print(f"Part 2 for puzzle input: {part2(input)}")

File: 2020/D19/d19.js
/*
Advent Of Code 2020
Day 19: Monster Messages part 1 & 2

https://adventofcode.com/2022/day/19
*/

const fs = require('fs');
const path = require('path');

const YEAR = '2020';
const DAY = '19';

function readInput(filename) {
  const readRawIntput = () => {
    const data = fs.readFileSync(path.join(`${YEAR}`, `D${DAY}`, filename));
    const lines = data
      .toString()
      .split(`\n`)
      .filter((l) => l);
    return lines;
  };

  const messages = [];
  const rules = [];
  const parseALine = (line) => {
    if (line.startsWith('a') || line.startsWith('b')) {
      messages.push(line);
      return;
    }

    const [id, rule] = line.split(':');
    const entry = {
      id: Number(id),
      resolved: false,
      rules: [[]],
    };
    rule
      .trim()
      .split(' ')
      .forEach((r) => {
        if (r === '|') entry.rules.push([]);
        else if (r === '"a"') {
          entry.rules[entry.rules.length - 1].push('a');
          entry.resolved = true;
        } else if (r === '"b"') {
          entry.rules[entry.rules.length - 1].push('b');
          entry.resolved = true;
        } else entry.rules[entry.rules.length - 1].push(Number(r));
      });

    rules.push(entry);
  };

  try {
    const lines = readRawIntput();
    lines.forEach((l) => parseALine(l));

    return { rules, messages };
  } catch (err) {
    console.error(err);
  }
}

function generatePermutations(list) {
  const result = [];
  const indexes = new Array(list.length).fill(0);
  const permutations = (level) => {
    if (level === indexes.length) {
      const r = list.reduce((r, item, index) => {
        r.push(item[indexes[index]]);
        return r;
      }, []);
      result.push(r);
      return;
    }
    while (indexes[level] < list[level].length) {
      permutations(level + 1);
      indexes[level] += 1;
    }
    indexes[level] = 0;
    return;
  };
  permutations(0);
  return result;
}
function resolveRule(rule1, rule2) {
  const { id } = rule1;

  let rulesToReduce = rule2.rules.filter((r) => r.some((v) => v === id));
  if (rulesToReduce.length === 0) return;
  console.log(`rule #:${rule1.id} -> #${rule2.id}`);
  const unchangedRules = rule2.rules.filter((r) => r.every((v) => v !== id));
  const newRuleSet = [...unchangedRules];

  // replace all rule that matched id by -1
  rulesToReduce = rulesToReduce.map((rule) => rule.map((v) => (v === id ? -1 : v)));
  // proceed with replacement of each matching rule
  rulesToReduce.forEach((rule) => {
    const reducedRule = rule.map((v) => (v === -1 ? rule1.rules.flat(1) : [v]));
    const p = generatePermutations(reducedRule);
    newRuleSet.push(...p);
  });

  // join rules that can be joined
  rule2.rules = newRuleSet.map((rule) => (rule.every((v) => !Number.isInteger(v)) ? [rule.join('')] : rule));

  if (rule2.rules.every((rule) => rule.every((v) => !Number.isInteger(v)))) {
    rule2.resolved = true;
    rule2.rules = rule2.rules.map((rule) => [rule.join('')]);
  }
  return rule2.resolved;
}

function part1(rules, messages) {
  rules.sort((a, b) => a.id - b.id);
  let resolvedRules = rules.filter(({ resolved }) => resolved);
  let remainingRules = rules.filter(({ resolved }) => !resolved);

  while (remainingRules.length > 0) {
    resolvedRules.forEach((rule1) => {
      remainingRules.forEach((rule2) => {
        resolveRule(rule1, rule2);
      });
    });
    resolvedRules = remainingRules.filter(({ resolved }) => resolved);
    remainingRules = remainingRules.filter(({ resolved }) => !resolved);
    console.log(`Resolve:${resolvedRules.length} Remaining:${remainingRules.length} `);
  }

  console.log('======================================');
  const ruleSet = new Set(resolvedRules[0].rules.map((r) => r[0]));
  const result = [...new Set(messages)].filter((m) => ruleSet.has(m));
  return result.length;
}

function part2(rules, messages) {
  let resolvedRules = rules.filter(({ resolved }) => resolved);
  let remainingRules = rules.filter(({ resolved }) => !resolved);

  const rule0 = rules.find(({ id }) => id === 0);
  const rule8 = rules.find(({ id }) => id === 8);
  const rule11 = rules.find(({ id }) => id === 11);

  //create loop in rule 8 and 11
  rule8.rules.push([42, 8]);
  rule11.rules.push([42, 11, 31]);

  while (resolvedRules.length > 1) {
    resolvedRules.forEach((rule1) => {
      remainingRules.forEach((rule2) => {
        resolveRule(rule1, rule2);
      });
    });
    resolvedRules = remainingRules.filter(({ resolved }) => resolved);
    remainingRules = remainingRules.filter(({ resolved }) => !resolved);
    console.log(`Resolve:${resolvedRules.length} Remaining:${remainingRules.length} `);
  }
  console.log('======================================');

  // we should be left with 3 rules ...
  // first partialy resolve rule 8 & 11
  rule8.rules = rule8.rules.map((rule) => {
    if (rule.every((s) => !Number.isInteger(s))) return [rule.join('')];
    return rule;
  });
  rule11.rules = rule11.rules.map((rule) => {
    if (rule.every((s) => !Number.isInteger(s))) return [rule.join('')];
    return rule;
  });

  const matchedMessages = [];
  while (true) {
    resolveRule(rule8, rule0);
    resolveRule(rule11, rule0);
    rule0.rules = rule0.rules.map((rule) => {
      if (rule.every((s) => !Number.isInteger(s))) return [rule.join('')];
      return rule;
    });
    const ruleSet = new Set(rule0.rules.filter((rule) => rule.every((r) => !Number.isInteger(r))).map((r) => r[0]));
    const result = [...new Set(messages)].filter((m) => ruleSet.has(m));
    if (result.length === 0) break;
    matchedMessages.push(...result);
    // remove resulved rules from rule 0
    rule0.rules = rule0.rules.filter((r) => r.some((v) => Number.isInteger(v)));
  }

  return matchedMessages.length;
}

let input = readInput(`d${DAY}-input.txt`);
let sample1 = readInput(`d${DAY}-sample1.txt`);
let sample2 = readInput(`d${DAY}-sample2.txt`);
console.log(`Part 1 sample1 : ${part1(sample1.rules, sample1.messages)}`);
console.log(`Part 1 sample2 : ${part1(sample2.rules, sample2.messages)}`);
console.log(`Part 1 input: ${part1(input.rules, input.messages)}`);

input = readInput(`d${DAY}-input.txt`);
sample2 = readInput(`d${DAY}-sample2.txt`);
console.log(`Part 2 sample2 : ${part2(sample2.rules, sample2.messages)}`);
console.log(`Part 2 input: ${part2(input.rules, input.messages)}`);
